compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs

```diff
@@ -2397,24 +2398,112 @@ impl<'a, 'tcx> InferCtxtExt<'tcx> for InferCtxt<'a, 'tcx> {
                     {
                         let parent_trait_ref =
                             self.resolve_vars_if_possible(data.parent_trait_pred);
-                        let ty = parent_trait_ref.skip_binder().self_ty();
-                        matches!(ty.kind(), ty::Generator(..))
-                            || matches!(ty.kind(), ty::Closure(..))
+                        let nested_ty = parent_trait_ref.skip_binder().self_ty();
+                        matches!(nested_ty.kind(), ty::Generator(..))
+                            || matches!(nested_ty.kind(), ty::Closure(..))
                     } else {
                         false
                     }
                 };
 
+                let future_trait = self.tcx.lang_items().future_trait().unwrap();
+                let opaque_ty_is_future = |def_id| {
+                    self.tcx.explicit_item_bounds(def_id).iter().any(|(predicate, _)| {
+                        if let ty::PredicateKind::Trait(trait_predicate) =
+                            predicate.kind().skip_binder()
+                        {
+                            trait_predicate.trait_ref.def_id == future_trait
+                        } else {
+                            false
+                        }
+                    })
+                };
+
+                let opaque_ty_is_core_from_generator = |def_id| {
+                    // HACK: this is ... pretty bad. Unfortunately, we can't tag this future with `rustc_diagnostic_item`
+                    // because it's not actually `GenFuture`, it's the opaque type itself.
+                    let source_map = tcx.sess.source_map();
+                    let file = |def_id| source_map.span_to_filename(tcx.def_span(def_id));
+
+                    let from_generator = tcx.lang_items().from_generator_fn().unwrap();
+                    file(def_id) == file(from_generator)
```

Comment:
i think there should be some relationship between opaque type and the function because this is an RPIT, right? what's the [parent def id](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.DefIdTree.html#tymethod.opt_parent) of the opaque type of an RPIT? it should be the function i hope

Reply:
ah, you're exactly right :) I'll change this, thanks!
```
DEBUG rustc_trait_selection::traits::error_reporting::suggestions note_obligation_cause_code: check core::from_generator, parent=DefId(2:13721 ~ core[5b88]::future::from_generator), def_id=DefId(2:13723 ~ core[5b88]::future::from_generator::{opaque#0})
```